### PR TITLE
onboarding screen added

### LINKED ISSUE #421: Adding a tutorial screen that includes app icon and instructions.
**Is it a ?**
 Feature Request
capture

  
  
#### Feature Description

For helping the first time users, the app should have a set of instructions that tell them what this app is all about and how to use the app along with the icon.

##Are you willing to work on the issue?

Yes.

### PR DESCRIPTION
**Pull request by harshvardhan-takawale**
Fixes #421 

Changes:
Introduction screens added in viewpager.

### CODE REVIEW

**Comment by harshvardhan-takawale** [comment]
@brijeshshah13 @mhutti1 Please review. I have created simple viewpager type of onboarding activity.
I don't know why Travis CI is taking so long.

**Comment by harshvardhan-takawale** [comment]
capture1
capture2
capture3
capture4
capture5

**Comment by srv-twry** [comment]
@harshvardhan-takawale The UI looks good but can you please attach a GIF instead of these screenshots in order to get a better feel of what is happening. Thanks.

**Comment by harshvardhan-takawale** [comment]
@srv-twry Sure. Here you go.

**Comment by mhutti1** [comment]
I would put skip in the middle and use arrows/next previous either side of it.

**Comment by harshvardhan-takawale** [comment]
@brijeshshah13 @mhutti1 I have added the "previous button" and changed the code according to proper indentation. I came to know now that there was a short-cut for doing this.

**Comment by harshvardhan-takawale** [comment]
@mhutti1 @brijeshshah13 @srv-twry any more suggestions?

**Comment by brijeshshah13** [comment]
@mhutti1 LGTM. Verified on API 25 device and API 16 emulator.

File: app/src/main/java/org/kiwix/kiwixmobile/KiwixMobileActivity.java [hunk 1]
```diff
@@ -429,6 +429,17 @@ public void onCreate(Bundle savedInstanceState) {
 
     ActionBarDrawerToggle drawerToggle = new KiwixActionBarDrawerToggle(this, drawerLayout, toolbar);
 
+    Boolean isFirstRun = getSharedPreferences("PREFERENCE", MODE_PRIVATE)
+            .getBoolean("isFirstRun", true);
```
Comment (mhutti1) [inline app/src/main/java/org/kiwix/kiwixmobile/KiwixMobileActivity.java]: 4 Space line continuations

File: app/src/main/java/org/kiwix/kiwixmobile/WelcomeActivity.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        if (current >= 0) {
+          viewPager.setCurrentItem(current);
+        }
+      }
+    });
+
+  }
+
+  private void addBottomDots(int position) {
+
```
Comment (mhutti1) [inline app/src/main/java/org/kiwix/kiwixmobile/WelcomeActivity.java]: no random lines like this

File: app/src/main/res/layout/welcome_detail1.xml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  android:orientation="vertical">
+
+    <LinearLayout
+      android:layout_width="wrap_content"
+      android:layout_height="wrap_content"
+      android:layout_centerInParent="true"
+      android:gravity="center_horizontal"
+      android:orientation="vertical">
+
+        <ImageView
```
Comment (mhutti1) [inline app/src/main/res/layout/welcome_detail1.xml]: If im counting right this is 4 spaces.

**Comment by harshvardhan-takawale** [comment]
@mhutti1 I apologize for repeating these mistakes. Please review.

File: app/src/main/java/org/kiwix/kiwixmobile/KiwixMobileActivity.java [hunk 4]
```diff
@@ -429,6 +429,17 @@ public void onCreate(Bundle savedInstanceState) {
 
     ActionBarDrawerToggle drawerToggle = new KiwixActionBarDrawerToggle(this, drawerLayout, toolbar);
 
+    Boolean isFirstRun = getSharedPreferences("PREFERENCE", MODE_PRIVATE)
```
Comment (mhutti1) [inline app/src/main/java/org/kiwix/kiwixmobile/KiwixMobileActivity.java]: We have a new shared preference system now.

File: app/src/main/java/org/kiwix/kiwixmobile/KiwixMobileActivity.java [hunk 5]
```diff
@@ -429,6 +429,17 @@ public void onCreate(Bundle savedInstanceState) {
...
+        .getBoolean("isFirstRun", true);
+
+    if (isFirstRun) {
+      startActivity(new Intent(KiwixMobileActivity.this, WelcomeActivity.class));
+    }
+
+
+    getSharedPreferences("PREFERENCE", MODE_PRIVATE).edit()
+            .putBoolean("isFirstRun", false).commit();
```
Comment (mhutti1) [inline app/src/main/java/org/kiwix/kiwixmobile/KiwixMobileActivity.java]: 4 space tab.

File: app/src/main/java/org/kiwix/kiwixmobile/WelcomeActivity.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    prev.setOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View view) {
+        int current = getItem(-1);
+        if (current >= 0) {
+          viewPager.setCurrentItem(current);
+        }
+      }
+    });
+
```
Comment (mhutti1) [inline app/src/main/java/org/kiwix/kiwixmobile/WelcomeActivity.java]: Please remove random whitespace such as this.

File: app/src/main/java/org/kiwix/kiwixmobile/WelcomeActivity.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+  ViewPager.OnPageChangeListener viewListener = new ViewPager.OnPageChangeListener() {
+    @Override
+    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
+
+    }
+
+    @Override
+    public void onPageSelected(int position) {
+
```
Comment (mhutti1) [inline app/src/main/java/org/kiwix/kiwixmobile/WelcomeActivity.java]: And this.

File: app/src/main/java/org/kiwix/kiwixmobile/KiwixMobileActivity.java (hunk 4 above)
Comment (mhutti1) [reply app/src/main/java/org/kiwix/kiwixmobile/KiwixMobileActivity.java]: actually my bad about this. It isn't in just yet. Keeping track of 25 pull requests is getting a bit out of hand :P

**Comment by harshvardhan-takawale** [comment]
@mhutti1 I have made the changes and rebased.

**Comment by harshvardhan-takawale** [comment]
@mhutti1 Any reviews for this?

**Comment by srv-twry** [comment]
@harshvardhan-takawale We now have a SharedPreferenceManager for managing default shared preferences. Please use that :smile:

**Comment by harshvardhan-takawale** [comment]
@srv-twry Done.

File: app/src/main/java/org/kiwix/kiwixmobile/KiwixMobileActivity.java [hunk 8]
```diff
@@ -425,6 +426,15 @@ public void onCreate(Bundle savedInstanceState) {
 
     ActionBarDrawerToggle drawerToggle = new KiwixActionBarDrawerToggle(this, drawerLayout, toolbar);
 
+    SharedPreferences firstRunPref = PreferenceManager.getDefaultSharedPreferences(this);
```
Comment (srv-twry) [inline app/src/main/java/org/kiwix/kiwixmobile/KiwixMobileActivity.java]: Please check the newly created ```SharedPreferenceManager``` and use it.

File: app/src/main/java/org/kiwix/kiwixmobile/WelcomeActivity.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    super.onCreate(savedInstanceState);
+
+    setContentView(R.layout.welcome_main);
+    viewPager = (ViewPager) findViewById(R.id.view_pager);
+    skip = (Button) findViewById(R.id.btn_skip);
+    next = (Button) findViewById(R.id.btn_next);
+    prev = (Button) findViewById(R.id.btn_prev);
+    prev.setVisibility(View.GONE);
+
+
```
Comment (srv-twry) [inline app/src/main/java/org/kiwix/kiwixmobile/WelcomeActivity.java]: Don't add too many extra lines.

File: app/src/main/java/org/kiwix/kiwixmobile/WelcomeActivity.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        } else {
+          Intent i = new Intent(WelcomeActivity.this, KiwixMobileActivity.class);
+          i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
+          startActivity(i);
+          finish();
+        }
+      }
+    });
+
+    prev.setOnClickListener(new View.OnClickListener() {
```
Comment (srv-twry) [inline app/src/main/java/org/kiwix/kiwixmobile/WelcomeActivity.java]: Let's try to keep ```onCreate``` short. Please create another method ```setOnClickListeners``` and extract this code from here and just call it from the ```onCreate```.

File: app/src/main/java/org/kiwix/kiwixmobile/WelcomeActivity.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+      } else {
+        next.setText(R.string.next);
+        skip.setVisibility(View.VISIBLE);
+        prev.setVisibility(View.VISIBLE);
+      }
+    }
+
+    @Override
+    public void onPageScrollStateChanged(int state) {
+
```
Comment (srv-twry) [inline app/src/main/java/org/kiwix/kiwixmobile/WelcomeActivity.java]: If there is nothing to do, then please comment either ```// No implementation``` or ```// Nothing to do```

File: app/src/main/java/org/kiwix/kiwixmobile/WelcomeActivity.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    }
+  };
+
+
+  public class ViewPagerAdapter extends PagerAdapter {
```
Comment (srv-twry) [inline app/src/main/java/org/kiwix/kiwixmobile/WelcomeActivity.java]: Can you please try and move it to a new file ? It helps keeps the code organised and activities shorter.

File: app/src/main/res/layout/welcome_detail1.xml [hunk 13]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (srv-twry) [inline app/src/main/res/layout/welcome_detail1.xml]: ```RelativeLayout``` is outdated in the favour of ```ConstraintLayout```. ConstraintLayouts are extremely efficient in performance. If possible, convert it to ConstraintLayout.

File: app/src/main/java/org/kiwix/kiwixmobile/KiwixMobileActivity.java (hunk 8 above)
Comment (harshvardhan-takawale) [reply app/src/main/java/org/kiwix/kiwixmobile/KiwixMobileActivity.java]: SharedPreferenceUtil?

Comment (srv-twry) [reply app/src/main/java/org/kiwix/kiwixmobile/KiwixMobileActivity.java]: Yes. Sorry for late reply

**Comment by harshvardhan-takawale** [comment]
@srv-twry @mhutti1 Please review.

**Comment by ISNIT0** [comment]
Nice stuff Harshvardhan.
My one point is that the icons/words shown in the introduction don't seem to map to the actual behaviour in the app.
E.g. "Get Some Knowledge":
- The icon isn't used in the actual app/on the landing page
- The words aren't the same as in the app
So for that specific example, it may be worth adding some text underneath saying "Tap Get Content to start downloading" or something?

- "Get files for different fields" doesn't really make a huge amount of sense - what do you mean here?

Just some thoughts, but it's looking great!

**Comment by harshvardhan-takawale** [comment]
@ISNIT0 Thanks! Actually, when I made this, I kept the text and the images as a placeholder and thought I would change it according to the reviews of others. Also, are there any other things that you would like to change in the functionality?

**Comment by ISNIT0** [comment]
Nothing in particular, but it may be worth posting a new GIF since your button changes.

**Comment by harshvardhan-takawale** [comment]
@ISNIT0 Any suggestion for the images I should use?

**Comment by ISNIT0** [comment]
@harshvardhan-takawale I would recommend only images/icons actually within the app (you could always add the to the app too).

E.g. the download page, could have that cloud-dl icon, and you could add the same to the "Get Content" button.

**Comment by brijeshshah13** [comment]
@harshvardhan-takawale Please post a new GIF of the onboarding screen.

**Comment by abdulwd** [comment]
Moved to #754

**Pull request closed without merging**